### PR TITLE
Modulo 4

### PR DESCRIPTION
**Pull request by Riemann-def**
1. Non-linear Models exploration
2. Production code

Currently we allow different model parametrizations in the fit handler, but we're using a fixed threshold (0.3039) from config. Should we also optimize and store the threshold for each new model parametrization, or is it better to keep the threshold fixed across different model configs? What's the recommended approach for production?"

### CODE REVIEW

File: src/module_4/config.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+from pathlib import Path
```
Comment (sergiorozada12) [inline src/module_4/config.py:1]: Great job here adding a config file. Consider using config classes to contain your params, that allows you to add explanations and docstrings too. Take a look at hydra or omegaconf.

File: src/module_4/predict.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import json
+from src.module_4.utils import validate_event, get_users_df
+from src.module_4.push_model import PushModel
+
+
+def handler_predict(event, _):
```
Comment (sergiorozada12) [inline src/module_4/predict.py:6]: Remember to add typings always.

File: src/module_4/predict.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+        preds = model.predict(users_df)
+        predictions = dict(zip(users_df.index.astype(str), preds.astype(int)))
+
+        return {
+            "statusCode": 200,
+            "body": json.dumps({"prediction": predictions}),
+        }
+
+    except ValueError as ve:
```
Comment (sergiorozada12) [inline src/module_4/predict.py:22]: Good job here catching the exceptions.

File: src/module_4/push_model.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import glob
+import joblib
+import pandas as pd
+from lightgbm import LGBMClassifier
+from sklearn.metrics import fbeta_score
+
+from src.module_4.config import THRESHOLD, MODEL_DIR, TODAY_STR
+
+
+class PushModel:
```
Comment (sergiorozada12) [inline src/module_4/push_model.py:12]: I like the idea of the generic class to contain the model. Can you thinnk of ways of decoupling this class from LGBMClassifier? That would be nice in a pro env, since this class allows you to interface the generic model you train with production in such a way later on you only have to adapt a new model to this class instead of the rest of the production code that depends on this piece.

**Review by sergiorozada12 — APPROVED** [review]
Good job and good PR. Merge whenever you want. As always, I leave you here some comments:
* Remember to also add typings to functions in the NBs. Often times you will end up adding them to libraries or pro envs.
* From a modeling point of view, very good analysis. You have been thorough at comparing models, you used a baselines, temporal-based splits, etc. Well done. Regarding your question with respect to the threshold. A threshold is a best hyperparameter for a given model. So of course if you change the model you should change the threshold accordingly to maximize whatever metric you choose.
* Now, I prefer leaving you here some food for thought about what else you can do about this if you think from a company / business perspective. For example, you said that the best model is the GBM, of course, because it gives you the best metric. It is true, but, is it way better than the alternatives? If performances are comparable, would you really go with GBM to production? What are the latencies of running the models? Easiness to adapt to code in pro? Dependencies that you generate? Etc. For example, if your business is latency constrained, you probably want to go with a linear model to pro if you are not loosing too much performance, since it's going to run way faster. On the other hand, it is clear by your feature analysis that the most important feature is ordered_before. Does that mean that there is a different regime for usual costumers and new costumers? Maybe it is worth it exploring things separately and you can do way better with usual customers while global popularity is a simple heuristic for new customers. Same applies for other data partitions. Maybe modeling products separately lead to a more insightful analysis.
